utils/env_preferences/src/apple.rs

```diff
@@ -2,135 +2,203 @@
-        string::CFStringGetCStringPtr,
-    };
-    use std::ffi::CStr;
-
-    pub fn get_locales() -> Vec<String> {
-        let mut languages: Vec<String> = Vec::new();
-
-        // SAFETY: The call to `CFLocaleCopyPreferredLanguages` returns an immutable reference to `CFArray` which is owned by us
-        // https://developer.apple.com/documentation/corefoundation/cfarrayref. It is ensured that `locale_carr_ref` is not mutated
-        // Immutablility ensures that nothing is overriden during it's scope
-        let locale_carr_ref = unsafe { CFLocaleCopyPreferredLanguages() };
-
-        if !locale_carr_ref.is_null() {
-            // SAFETY: The call to `CFArrayGetCount` is only made when is `locale_carr_ref` is not `NULL`
-            let count = unsafe { CFArrayGetCount(locale_carr_ref as _) };
-            for i in 0..count {
-                // SAFETY: The call to `CFArrayGetValueAtIndex` is safe because we are iterating from 0 to count (`CFArrayGetCount`) which
-                // in itself will always be greater than 0 and less than count ensuring we will not get "out of bounds" error
-                let lang_ptr = unsafe { CFArrayGetValueAtIndex(locale_carr_ref as _, i) };
-
-                if !lang_ptr.is_null() {
-                    // SAFETY: The call to `CFStringGetCStringPtr` because the reference of string we are accessing is not `NULL`
-                    // Returns pointer in O(1) without any memory allocation. This can return NULL so to handle it
-                    let lang_str = unsafe { CFStringGetCStringPtr(lang_ptr as _, 0) };
-
-                    if !lang_str.is_null() {
-                        // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
-                        let lang_rust_str = unsafe { CStr::from_ptr(lang_str) }
-                            .to_str()
-                            .unwrap_or("Unknown")
-                            .to_string();
-                        languages.push(lang_rust_str);
-                    }
-
-                    // TODO: In case `lang_str` is `NULL` try retrieving the string using `CFStringGetCString`
-                    // Ref: https://developer.apple.com/documentation/corefoundation/1542721-cfstringgetcstring?language=objc
-                    // Note: Not optimal and may give inconsistent results if buffer is not large enough, must add sanity checks
-                    // whenever implemented
+use core_foundation_sys::{
+    array::{CFArrayGetCount, CFArrayGetValueAtIndex},
+    base::{CFIndex, CFRelease, CFRetain},
+    calendar::{CFCalendarCopyCurrent, CFCalendarCopyLocale, CFCalendarGetIdentifier},
+    locale::{CFLocaleCopyPreferredLanguages, CFLocaleGetIdentifier},
+    string::{
+        kCFStringEncodingUTF8, CFStringGetCString, CFStringGetCStringPtr, CFStringGetLength,
+        CFStringRef,
+    },
+    timezone,
+};
+use libc::c_char;
+use std::ffi::CStr;
+
+use crate::RetrievalError;
+
+// Helper function
```

Comment:
I'm confused by what this actually does. In other places you do

```rust
CStr::from_ptr(
    CFStringGetCStringPtr(
        ptr as CFStringRef, 
        kCFStringEncodingUTF8
    )
)
.to_str()?
.to_string()
```

which seems to do the same without manual allocations.

Reply:
Actually, this is a first priority to do without manual allocations. I had test failing in CI (not locally). Later realised this on apple [website](https://developer.apple.com/documentation/corefoundation/1542133-cfstringgetcstringptr?language=objc):

```
This function either returns the requested pointer immediately, 
with no memory allocations and no copying, 
in constant time, or returns NULL. If the latter is the result, call 
an alternative function such as the 
CFStringGetCString function to extract the characters.
```

Reply:
Please encapsulate the whole logic, including the null check in this function then, and add the comment explaining what you're doing. `get_string` doesn't tell me anything.